adapter/internal/adapter/adapter.go

```diff
@@ -0,0 +1,205 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+// Package adapter contains the implementation to start the adapter
+package adapter
+
+import (
+	"crypto/tls"
+	"time"
+
+	enforcerCallbacks "github.com/wso2/apk/adapter/internal/discovery/xds/enforcercallbacks"
+	apiservice "github.com/wso2/apk/adapter/pkg/discovery/api/wso2/discovery/service/api"
+	configservice "github.com/wso2/apk/adapter/pkg/discovery/api/wso2/discovery/service/config"
+	subscriptionservice "github.com/wso2/apk/adapter/pkg/discovery/api/wso2/discovery/service/subscription"
+	wso2_server "github.com/wso2/apk/adapter/pkg/discovery/protocol/server/v3"
+	"github.com/wso2/apk/adapter/pkg/logging"
+	"github.com/wso2/apk/adapter/pkg/tlsutils"
+
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"os/signal"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/wso2/apk/adapter/config"
+	"github.com/wso2/apk/adapter/internal/discovery/xds"
+	logger "github.com/wso2/apk/adapter/internal/loggers"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+var (
+	debug       bool
+	onlyLogging bool
+
+	localhost = "0.0.0.0"
+
+	port        uint
+	gatewayPort uint
+	alsPort     uint
+
+	mode string
+)
+
+const (
+	ads          = "ads"
+	amqpProtocol = "amqp"
+)
+
+func init() {
+	flag.BoolVar(&debug, "debug", true, "Use debug logging")
+	flag.BoolVar(&onlyLogging, "onlyLogging", false, "Only demo AccessLogging Service")
+	flag.UintVar(&port, "port", 18000, "Management server port")
+	flag.UintVar(&gatewayPort, "gateway", 18001, "Management server port for HTTP gateway")
+	flag.UintVar(&alsPort, "als", 18090, "Accesslog server port")
+	flag.StringVar(&mode, "ads", ads, "Management server type (ads, xds, rest)")
+}
+
+const grpcMaxConcurrentStreams = 1000000
+
+func runManagementServer(conf *config.Config, enforcerServer wso2_server.Server, enforcerSdsServer wso2_server.Server,
+	enforcerAppDsSrv wso2_server.Server, enforcerAppKeyMappingDsSrv wso2_server.Server, port uint) {
+	var grpcOptions []grpc.ServerOption
+	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
+	publicKeyLocation, privateKeyLocation, truststoreLocation := tlsutils.GetKeyLocations()
+	cert, err := tlsutils.GetServerCertificate(publicKeyLocation, privateKeyLocation)
+
+	caCertPool := tlsutils.GetTrustedCertPool(truststoreLocation)
+
+	if err == nil {
+		grpcOptions = append(grpcOptions, grpc.Creds(
+			credentials.NewTLS(&tls.Config{
+				Certificates: []tls.Certificate{cert},
+				ClientAuth:   tls.RequireAndVerifyClientCert,
+				ClientCAs:    caCertPool,
+			}),
+		))
+	} else {
+		logger.LoggerMgw.Warn("failed to initiate the ssl context: ", err)
+		panic(err)
```

Comment:
Feel it's better to add a blocking error log instead

Reply:
changed to Error